src/main/java/net/fabricmc/loader/launch/knot/Knot.java

```diff
@@ -0,0 +1,433 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.launch.knot;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.loader.launch.common.FabricLauncherBase;
+import net.fabricmc.loader.launch.common.FabricMixinBootstrap;
+import net.fabricmc.loader.launch.common.MixinLoader;
+import net.fabricmc.loader.transformer.PublicAccessTransformer;
+import org.spongepowered.asm.launch.MixinBootstrap;
+import org.spongepowered.asm.mixin.MixinEnvironment;
+import org.spongepowered.asm.mixin.transformer.MixinTransformer;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.*;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+
+public final class Knot extends FabricLauncherBase {
+	protected Map<String, Object> properties = new HashMap<>();
+
+	private PatchingClassLoader loader;
+	private boolean isDevelopment;
+	private EnvType envType;
+	private String entryPoint;
+	private File gameJarFile;
+	private List<URL> classpath;
+
+	private static class NullClassLoader extends ClassLoader {
+		private static final Enumeration<URL> NULL_ENUMERATION = new Enumeration<URL>() {
+			@Override
+			public boolean hasMoreElements() {
+				return false;
+			}
+
+			@Override
+			public URL nextElement() {
+				return null;
+			}
+		};
+
+		static {
+			registerAsParallelCapable();
+		}
+
+		@Override
+		protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+			throw new ClassNotFoundException(name);
+		}
+
+		@Override
+		public URL getResource(String name) {
+			return null;
+		}
+
+		@Override
+		public Enumeration<URL> getResources(String var1) throws IOException {
+			return NULL_ENUMERATION;
+		}
+	}
+
+	private static class DynamicURLClassLoader extends URLClassLoader {
+		public DynamicURLClassLoader(URL[] urls) {
+			super(urls, new NullClassLoader());
+		}
+
+		public void addURL(URL url) {
+			super.addURL(url);
+		}
+
+		static {
+			registerAsParallelCapable();
+		}
+	}
+
+	private static class PatchingClassLoader extends ClassLoader {
```

Comment:
I've not been able to go through things in much detail yet (sorry), but it feels like it might be worth moving some of the larger intermediate class loaders into separate files.

Reply:
These are specific to Knot, and they're not really intermediate. They build a very delicate order of processing classes.